Implement noodlification

Adda0 opened the pull request:
This PR implements function `noodlify` for noodlification of NFA.

As agreed upon, we implement a naive noodlification method creating all noodles: all permutations of epsilon transitions, one from each depth. In the following PR, we will improve the current implementation with a new algorithm, creating only actually possible noodles which have a path in the original automaton. 

The code has been tested with some tests.

martinhruska commented:
Why is `noodlify.h` needed? Could not be the function declared in nfa.hh?

Adda0 commented:
It could be, of course. However, when discussing this with @vhavlena, we decided that it would be better for noodlify to have its own header and source files for the following reasons:
- The function stands somewhat aside from the normal operations on NFA. We think it would make sense to include a separate header file if the user wants to work with these operations concerning noodlification.
- We expect to have multiple new functions working with noodlification and its variations implementation in the future. For that reason, we think it is better to have a separate header file to explicitly distinguish between normal NFA operations and the operation for noodlification and related functions.

tfiedor commented on src/nfa/noodlify.cc:
```diff
@@ -0,0 +1,79 @@
+    Segmentation segmentation{ aut, epsilon };
+    const auto& epsilon_depths{ segmentation.get_epsilon_depths() };
+
+    Nfa noodle{}; // Noodle created for each combination of epsilon transitions.
```
What the fuck is noodle?! I would advise to add to the comment some brief explanation what is the noodle, whether it is something well-establish or something own :-)

I know, Lukas spoke about that, but I'm not exactly sure what exactly it is.

Adda0 replied:
Good idea. As I am working with them, I did not think about how someone else could have problems deciphering the meaning of noodles at all. I will make sure to add sufficient explanations.

tfiedor replied:
Yes, that's called `curse of the knowledge`, you are familiar with the term, since you are working with it for quite some time, but for new users it could be confusing.

Adda0 replied:
There is an explanation of what noodle is in the declaration of `noodlify()`. And `noodlify()` then only creates all possible noodles, as written in the documentation for `noodlify()`. I think anyone trying to use `noodlify()` will read the documentation for the function. Afterwards, they will know what a noodle is and what will `noodlify()` do.

I think this explanation is enough to give our users a good enough understanding of the function.

tfiedor replied:
Yeah, sorry, didn't notice it in the comment. That was actually what I was expecting :-)

tfiedor commented on src/nfa/noodlify.cc:
```diff
@@ -0,0 +1,79 @@
+
+        noodle = aut;
+
+        for (int depth{ 0 }; depth < static_cast<int>(epsilon_depths.size()); ++depth)
```
Required: Extract the call to size() out of the loop. It can potentially lead to accidental quadratic complexity.

Adda0 replied:
Compilers are proficient in making these optimizations in for loop condition checks whenever possible. However, there is no need to have to rely on the compiler to make this optimization. I will refactor. Thank you for the suggestion.

tfiedor replied:
You're probably right, but I would not depend on the compiler :-)

tfiedor reviewed (approved):
I'm OK with the PR except for two minor things:

  1. Add to comments what the noodlification and noodle is.
  2. Fix the possible accidental quadratic complexity.

Adda0 commented:
I have addressed the mentioned issues. Furthemore, I have split the noodlify function into smaller functions in hopes to reuse the smaller functions in the future for aforementioned additional operations for noodlification.
